pkg/api/v20191231preview/openshiftcluster_validate_test.go

```diff
@@ -0,0 +1,229 @@
+			f: func(change *OpenShiftCluster) {
+				change.Properties.WorkerProfiles = append(change.Properties.WorkerProfiles, WorkerProfile{
+					Name:       "worker",
+					VMSize:     VMSizeStandardD4sV3,
+					DiskSizeGB: 256,
+					SubnetID:   subnetID,
+					Count:      7,
+				})
+			},
+			wantErr: true,
+		},
+		{
+			name: "worker subnet change",
+			f: func(change *OpenShiftCluster) {
+				wp := change.Properties.WorkerProfiles[0]
+				wp.SubnetID = vnetID + "/subnets/other"
+				change.Properties.WorkerProfiles = []WorkerProfile{wp}
+			},
+			wantErr: true,
+		},
+		{
+			name: "count can change",
+			f:    func(change *OpenShiftCluster) { change.Properties.WorkerProfiles[0].Count = 15 },
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			oc := *old
+			tt.f(&oc)
+			if err := oc.validateDelta(old); (err != nil) != tt.wantErr {
+				t.Errorf("OpenShiftCluster.validateDelta() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
+
+func mutateAndTestStructField(t *testing.T, inOc, origOc *OpenShiftCluster, in, orig reflect.Value, inField reflect.StructField) {
+	immutable := inField.Tag.Get("mutable") != "true"
+	switch in.Kind() {
+	case reflect.Slice:
+		mutateAndTest(t, inOc, origOc, in, orig)
+	case reflect.Interface:
+		mutateAndTest(t, inOc, origOc, in, orig)
+	case reflect.Map:
+		mutateAndTest(t, inOc, origOc, in, orig)
+	case reflect.Array:
+		mutateAndTest(t, inOc, origOc, in, orig)
+	case reflect.Ptr:
+		mutateAndTest(t, inOc, origOc, in, orig)
+	case reflect.Struct:
+		mutateAndTest(t, inOc, origOc, in, orig)
+	case reflect.String:
+		in.SetString("test")
+		err := inOc.validateDelta(origOc)
+		if immutable == (err == nil) {
+			t.Errorf("%s immutable:%v, err %v", inField.Name, immutable, err)
+		}
+		in.Set(orig)
+	case reflect.Int:
+		inV := in.Interface().(int64)
+		in.SetInt(inV + 2)
+		err := inOc.validateDelta(origOc)
+		if immutable == (err == nil) {
+			t.Errorf("%s immutable:%v, err %v", inField.Name, immutable, err)
+		}
+		in.Set(orig)
+	case reflect.Bool:
+		inV := in.Interface().(bool)
+		in.SetBool(!inV)
+		err := inOc.validateDelta(origOc)
+		if immutable == (err == nil) {
+			t.Errorf("%s immutable:%v, err %v", inField.Name, immutable, err)
+		}
+		in.Set(orig)
+	default:
+	}
+}
+
+func mutateAndTest(t *testing.T, inOc, origOc *OpenShiftCluster, in, orig reflect.Value) {
+	switch in.Kind() {
+	case reflect.Array:
+		for i := 0; i < in.Len(); i++ {
+			mutateAndTest(t, inOc, origOc, in.Index(i), orig.Index(i))
+		}
+	case reflect.Interface:
+		if in.IsNil() || orig.IsNil() {
+			return
+		}
+		mutateAndTest(t, inOc, origOc, in.Elem(), orig.Elem())
+	case reflect.Map:
+		if in.IsNil() || orig.IsNil() {
+			return
+		}
+		for _, k := range in.MapKeys() {
+			mutateAndTest(t, inOc, origOc, in.MapIndex(k), orig.MapIndex(k))
+		}
+	case reflect.Ptr:
+		if in.IsNil() || orig.IsNil() {
+			return
+		}
+		mutateAndTest(t, inOc, origOc, in.Elem(), orig.Elem())
+	case reflect.Slice:
```

Comment:
so this is why the manual test is still in place above, any bright ideas?
I suspect i'll have to totally restructure the flow to account for this